File: 2020/11.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "input_11.txt"
    lines = readInput(filename)
    seats = Dict{Complex{Int}, Vector{Int}}()
    for y in eachindex(lines)
        for x in eachindex(lines[1])
            if lines[y][x] == 'L'
                seats[Complex(x, y)] = [0, 0]
            end
        end
    end

    return seats
end

function part1(seats::Dict{Complex{Int}, Vector{Int}})
    dirs = [1+1im, 1+0im, 1-1im,
            +1im, -1im,
            -1+1im, -1+0im, -1-1im]
    while true
        changes = false
        for (k, v) in seats

            neighs = 0
            for d in dirs
                neighs += haskey(seats, k+d) ? seats[k+d][1] : 0
            end
            if neighs == 0 && v[1] == 0
                seats[k][2] = 1
                changes = true
            elseif v[1] == 1 && neighs >= 4
                seats[k][2] = 0
                changes = true
            end
        end
    
        for k in keys(seats)
            seats[k][1] = seats[k][2]
        end

        #drawSeats(seats)
        if !changes
            break
        end
    end
    println("🎄 Part 1: $(sum([v[1] for v in values(seats)]))")
end

function drawSeats(seats::Dict{Complex{Int}, Vector{Int}})
    xmax = maximum([real(k) for k in keys(seats)])
    ymax = maximum([imag(k) for k in keys(seats)])
  
    for y in 1:ymax
        for x in 1:xmax
            if haskey(seats, Complex(x, y))
                if seats[Complex(x, y)][1] == 1
                    print("#")
                elseif seats[Complex(x, y)][1] == 0
                    print("L")
                end
            else
                print(".")
            end
        end
        println()
    end
    println()
end

function part2(seats::Dict{Complex{Int}, Vector{Int}})
    xmax = maximum([real(k) for k in keys(seats)])
    ymax = maximum([imag(k) for k in keys(seats)])
    
    dirs = [1+1im, 1+0im, 1-1im,
            +1im, -1im,
            -1+1im, -1+0im, -1-1im]
    while true
        changes = false
        for (k, v) in seats
            neighs = 0
            for d in dirs
                pos = k
                while true
                    pos += d
                    if !(1 <= real(pos) <= xmax && 1 <= imag(pos) <= ymax)
                        break
                    end
                    if haskey(seats, pos)
                        if seats[pos][1] == 1
                            neighs += 1
                        end
                        break
                    end
                end
            end

            if neighs == 0 && v[1] == 0
                seats[k][2] = 1
                changes = true
            elseif v[1] == 1 && neighs >= 5
                seats[k][2] = 0
                changes = true
            end
        end

        for k in keys(seats)
            seats[k][1] = seats[k][2]
        end
        #drawSeats(seats)
        
        if !changes
            break
        end
    end

    println("🎄🎅 Part 2: $(sum([v[1] for v in values(seats)]))")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 11        ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()
@time part1(inputs)
inputs = loadInput()
@time part2(inputs)

File: 2019/11.py
import time

from queue import Queue
from threading import Thread

from utils import readInput
from intcode import IntCode
            
def loadInput():
    return readInput("input_11.txt")

def painting(prog, hull, channel):
    thread = Thread(target=prog.run, daemon=True)
    thread.start()
    dir = 0
    pos = (0, 0)
    while True:
        for i in range(2):
            val = channel["me"].get()
            if val == "BYE":
                return
            if i == 0:
                hull[pos] = val
            else:
                if val == 0:
                    dir -= 1
                else:
                    dir += 1
                dir = dir % 4
        if dir == 0:
            pos = (pos[0], pos[1]-1)
        elif dir == 1:
            pos = (pos[0]+1, pos[1])
        elif dir == 2:
            pos = (pos[0], pos[1]+1)
        elif dir == 3:
            pos = (pos[0]-1, pos[1])
        channel["Robot"].put(hull.get(pos, 0))
    
def part1(lines):
    channel = {"Robot":Queue(), "me":Queue()}
    prog = IntCode("Robot", lines[0], qin=channel['Robot'], qout=channel['me'], mode="channel")
    channel['Robot'].put(0)
    hull = {}
    painting(prog, hull, channel)
    print (f"🎅 Part 1: {len(hull)}")

def show_hull(hull):
    xs = [h[0] for h in hull]
    ys = [h[1] for h in hull]
    xmin, xmax = min(xs), max(xs)
    ymin, ymax = min(ys), max(ys)
    for y in range(int(ymin), int(ymax+1)):
        for x in range(int(xmin), int(xmax+1)):
            val = hull.get((x, y), 0)
            if val == 1:
                print("▮", end='')
            else:
                print(" ", end='')
        print()
    
def part2(lines):
    channel = {"Robot":Queue(), "me":Queue()}
    channel['Robot'].put(1)
    prog = IntCode("Robot", lines[0], qin=channel['Robot'], qout=channel['me'], mode="channel")
    hull = {}
    painting(prog, hull, channel)
    print (f"🎅🎄 Part 2:")
    show_hull(hull)

if __name__ == "__main__":
    title = "Day 11: Space Police"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    lines = loadInput()
    
    t0 = time.time()
    part1(lines)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(lines)
    print ("Time: {:.5f}".format(time.time()-t0))
